Feature: unsubmitted mailer

jay-hankins opened the pull request:
Okay, so I've tried to set up a mailer to fire when a user is created. I'm using the letter-opener and launchy gems to emulate sending and receiving an email in browser (rather than actually against SendGrid) for now. I combined elements of both these tutorials, https://sendgrid.com/docs/Integrate/Frameworks/rubyonrails.html and https://github.com/Backgrounders/email-sidekiq-rails-4.2. To get this far.

It doesn't work. :disappointed: From console, it doesn't look like the UserNotifier class ever gets called through user create.

Paging Rails Dr. @kirbyk: Can you see anything off the bat as to why it's not working?

jay-hankins commented:
Which, this is just a preliminary effort before I get sidekiq working to check for unsubmitted timecards. Let's put the horse before the cart and get emails working first :+1:

kirbyk commented:
This looks like it should work. What specific errors are you getting? Feel free to paste the in the comments.

jay-hankins commented:
Okay. No error messages, just no emails being sent (displayed). 

Expected behavior: User goes to register from root page. User registers. User receives email confirming sign up (a new browser window pops up with the email).

Observed behavior: User registers and is returned a success message but no email is sent (not displayed). 

It works as expected in the rails console:
```
irb(main)> @u=User.new(first_name: "Beyonce", last_name:"Knowles", email:"beyonce@bey.com")
irb(main)> UserNotifier.send_signup_email(@u).deliver
```
Maybe @aloeb can look into this because it interfaces with the user registration model. It's as if the registration form doesn't hit user.save.

kirbyk commented:
So you get an email when you do it in the console...?!

If that's the case, then you should add `better-errors` to the Gemfile, and then add a `raise 'hell'` into the user controller right before the call to `UserNotifier`, and run the command in the REPL.

None of that might have made sense, but if you just do what I say in order it will make sense, I promise.

kirbyk commented:
This is the best way to debug.

jay-hankins commented:
Yes, email in console. Thanks for the debug tips, I'll look into it. in the long run, I'm not sure it matters. I just wanted to make sure I got the mailer working, which I did. Unless we want user confirmation emails?

jay-hankins commented:
Update: I added better_errors, added `raise 'hell'`,  ran `bundle-install`, and went to register a user again. Nothing different happened. I was expecting something like this after clicking the submit button... ![this](https://camo.githubusercontent.com/3fa6840d5e20236b4f768d6ed4b42421ba7c2f21/68747470733a2f2f692e696d6775722e636f6d2f367a42474141622e706e67). Where did I go wrong?

kirbyk commented:
You should raise hell right before the UserNotifier. This way in the REPL you can check out the @user variable, etc.

jay-hankins commented:
Getting closer. The UserNotifier should go into registrations_controller.rb.

kirbyk commented:
No, you did in in the index action. You should do it in the create action the line above UserNotifier.

What makes you say that? I think it should work in the user controller...

jay-hankins commented:
Actual snippet of my code:
```ruby
def create
    @user = User.new(user_params)

    respond_to do |format|
      if @user.save
        # Raise hell to access REPL and figure out why emails aren't being generated.
        raise 'hell' 
        # Deliver the signup email
        UserNotifier.send_signup_email(@user).deliver
        redirect_to(@user, :notice => 'User created')

        format.html { redirect_to @user, notice: 'User was successfully created.' }
        format.json { render :show, status: :created, location: @user }

      else
        raise 'hell'
        format.html { render :new }
        format.json { render json: @user.errors, status: :unprocessable_entity }
      end
    end
  end
```

kirbyk commented:
Okay, cool. That's good. You get a REPL their, right?

If you manually type in `UserNotifier.send_signup_email(@user).deliver` what do you get?

jay-hankins commented:
Check the latest commit. It works as expected if I raise hell and send_signup_email in /users/registrations_controller.rb. 

P.S. I've had raise hell in that place since yesterday... Really makes me think we aren't hitting the user controller, at least not through the registration page. This is probably something @aloeb did with devise that I'm not following.

kirbyk commented:
Ohhhh. That's totally what's happening. Registration controller is where it should be.

So everything is good to go then?

jay-hankins commented:
I believe so. This is to fire an email for user submit, obviously. Do we want this behavior? If yes, I'll branch it off and and do a separate PR. I still have to figure out how to make sidekiq check for unsubmitted cards.

kirbyk commented:
By User submit, you mean User registration, right?

Yeah, we should send them a verification email, but we should route that through Devise.

This can be used for a followup email though. Once they've verified, we can send them a nice email explaining our the system works and stuff.

jay-hankins commented:
Yes, user registration. I'll move this into a new branch for nice explanation emails.

kirbyk commented:
Okay, cool. Closing this PR.

jay-hankins commented:
I'm still trying to get an unsubmitted mailer working per #11. Shall we not delete this branch?

kirbyk commented:
Up to you. You can just play with it locally.

jay-hankins commented:
Alright, check out what I've got going on. I've got a sidekiq worker to send emails of unsubmitted timecards. Right now, it seems like timecards aren't correctly linked to users, but I tested it through the console and it works. 

I left some testing things in here, namely launchy, letter-opener, sinatra, and better_errors, just so we can see what's going on. Merge her in?

kirbyk commented:
Yeah, go ahead and merge.

Can you elaborate about not being linked..?

jay-hankins commented:
So it looks like the timecard submit doesn't set a user into a timecard. Just from looking at Rails docs, it looks like each timecard expects a user ID to be set into it, which as far as I can tell, isn't happening. When I did it in console, I just made a new user (@u), made a new timecard (@t), and then set @t.user =  @u. Then it worked as expected.

kirbyk commented:
Hmm. Can you check the controller, and add that in, then?

jay-hankins commented:
Yeah, looking into it. I'm not really following what @aloeb has done with the Timecard model. I suppose somehow we grab the user that's logged in (devise?) and then just push that into the Timecard.create method.

kirbyk commented:
Show me where you want clarification.

jay-hankins commented on app/workers/submit_checker_worker.rb:
```diff
@@ -0,0 +1,9 @@
+
+  def perform()
+    Timecard.where("status == 0").find_each do |timecard|
+        UserNotifier.send_signup_email(timecard.user).deliver
```
@kirbyk Okay, so right here is where I need to grab the user from a timecard.

kirbyk replied:
Does `timecard.user` == `nil`, or what?

jay-hankins replied:
Yes.

kirbyk replied:
Yeah, then adding that code I mentioned should fix it.

aloeb commented:
I think the user_id field is the implementation of the has_many in the User and the belongs_to in the Timecard. So I believe it functions as the link between an individual timecard and a user.

jay-hankins commented:
![Great success](https://coachhonea.files.wordpress.com/2011/01/great_success.png)

kirbyk commented:
NOW!!!
